### PR TITLE
Refactored Drop API

### LINKED ISSUE #247: API for sending a drop message / signing of drop message
I browsed through the test code of the drop component found some things
* first of all I find it really complicated to send a drop message:
I have to instantiate a drop message by my own and set several parameters (of which some don't depend on the senders state, e.g. the timestamp, version, acknowledgeID) before I can finally send it.
Is there a reason why that's not encapsulated somewhere (e.g. DropController)? Or is it just because of the last milestone's urgency?
* Additionally I wonder why the encrypted drop message gets signed with the sign key of the recipient instead of the sender (in class DropController:
```cryptedMessage = encryptDrop(serialize(message),
					contact.getEncryptionPublicKey(),
					contact.getContactOwner().getPrimaryKeyPair().getSignKeyPairs().get(0));```)
Or do I miss something here?
* The same happens when decrypting the message, the private key of a contact is used where the one of an identity should be used.

Edit:
Wait, now I see that the contact owner is used. so forget the last two points ;)

### PR DESCRIPTION
**Pull request by cburkert**
I refactored drop API with particular focus on
* use model classes instead of serialization representations wherever possible
* make DropMessage construction easier

This PR
* introduces a abstraction layer called Entity above Contact and Identity that contains common properties like drops, public key getters
* introduces an general mapper for Entity lookups by key identifier and uses it for Contacts and Identites
* implements acknowledge ID generation using UUIDv4
* implements detection of spoofed sender information in DropMessages
* complains about unexpected drop message version by throwing parser exception instead of treating the version field like a meaningless field

Closes #247 

EDIT: Added version checking "functionality" description

### CODE REVIEW

File: src/main/java/de/qabel/core/config/Contact.java [hunk 1]
```diff
@@ -170,18 +87,19 @@ public Contact(Identity owner)
 	 * Attention: This constructor is intended for deserialization purposes. The contactOwner needs to be set afterwards
 	 * @param ownerKeyId
```
Comment (aBothe) [inline src/main/java/de/qabel/core/config/Contact.java]: Thanks for the 'Attention' rationale in the comment! :)
The `@param`s for the other two parameters are missing.

File: src/main/java/de/qabel/core/config/EntityMap.java [hunk 2]
```diff
@@ -0,0 +1,68 @@
+package de.qabel.core.config;
+
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+
+abstract public class EntityMap<T extends Entity> {
```
Comment (aBothe) [inline src/main/java/de/qabel/core/config/EntityMap.java]: Despite I know that a map is some sort of lookup, it doesn't get clear what entities are mapped to (their keyIdentifiers?). Please mention this in the comment.

File: src/main/java/de/qabel/core/drop/DropDeserializer.java [hunk 3]
```diff
@@ -18,13 +18,15 @@
                                                 JsonDeserializationContext context) throws JsonParseException {
 
         int version          = json.getAsJsonObject().get("version")       .getAsInt();
+        if (version != DropMessage.getVersion()) {
+        	throw new JsonParseException("Unexpected version: " + version);
```
Comment (aBothe) [inline src/main/java/de/qabel/core/drop/DropDeserializer.java]: This new functionality should be mentioned in the PR description as well. Not sure whether the logic should at least try to examine readable info -- perhaps some versions are backward compatible or so. But well, not in this PR! :)

File: src/main/java/de/qabel/core/drop/DropMessage.java [hunk 4]
```diff
@@ -1,80 +1,102 @@
...
+    		return false;
+    	}
+    	this.sender = sender;
+    	return true;
     }
 
-    public void setModelObject(Class<T> modelObject) {
-        this.modelObject = modelObject;
+    public Class<T> getModelObject() {
+        return (Class<T>)data.getClass();
```
Comment (aBothe) [inline src/main/java/de/qabel/core/drop/DropMessage.java]: This seems way more elegant than explicitly storing some class name in an extra property.
Despite that, what if data happens to be `null`?

In my first Sync-approach, I'm handing in a `new DropMessage<SyncDropMessage>()` without changing anything else into DropController.sendAndForget() - so there will be no data set. I'm not sure whether this is sufficient though, but using DropMessage that way seems reasonable to me.

**Comment by aBothe** [comment]
In spite of the initial things I've critizied, the rest of this pr looks fine to me. Now let's place bets on how long it'll take for it to get merged.

**Comment by thechauffeur** [comment]
So this also closes #265 and also implements acknowledge ID generation and also implements spoofing detection.
Well... it would be good if you split your pull requests more often or write issues like "Implement spoofing detection" and assign them to yourself and wait with implementing them until merging (fixing) your current PR (if they depend on each other).
Your PRs are not really nice to review and also implement / fix stuff without a corresponding issue. Have we even decided on the acknowledge IDs yet? Also referencing all effected issues would be really really nice...

File: src/main/java/de/qabel/core/config/Contact.java (hunk 1 above)
Comment (cburkert) [inline src/main/java/de/qabel/core/config/Contact.java]: Well, I didn't write the attention note. But you're right, the other params are missing. I'll remove the one param comment because the are self-explaining and redundant anyway.

File: src/main/java/de/qabel/core/drop/DropDeserializer.java (hunk 3 above)
Comment (cburkert) [inline src/main/java/de/qabel/core/drop/DropDeserializer.java]: Well, I don't see much of a new functionality in this, but I'll add it to the description.

File: src/main/java/de/qabel/core/drop/DropMessage.java (hunk 4 above)
Comment (cburkert) [inline src/main/java/de/qabel/core/drop/DropMessage.java]: I don't see why you should pass null as payload instead of an instance of SyncDropMessage.

**Comment by cburkert** [comment]
@aBothe I cannot change the send API to Entity as recipient type that easy because, the Contact-specific contactOwner property is used to get the private keys for signing. I could use the sender reference in the DropMessage instead but have to think about the implications first.
I'd rather like to work this out in another PR, but I understand your thoughts.

**Comment by cburkert** [comment]
@thechauffeur I'm not sure if this PR closes #265 because this is not a feature request issue but a discussion issue and there are points in that discussion (e.g. about whether or not a Contact is a setting) which are not magically solved even if PR is eventually merged.

The acknowlege IDs are [documented in the spec](https://github.com/Qabel/qabel-doc/wiki/Qabel-Client-Drop#acknowledging). Since I had to return any acknowledge ID in the getter to avoid syntactical errors, I thought a one-liner that creates UUIDs is at least as good a solution as returning an empty string.

The spoofing detection is also a really insignificant by-product. Taking four LOCs it should not complicate this PR much.

I'm sorry that this PR is large, but since it is a refactoring PR, I couldn't split it up while maintaining any meaning.

**Comment by thechauffeur** [comment]
@cburkert OK, I see.

**Comment by cburkert** [comment]
Changed the serialization of the properties Contact inherits from SyncSettingItem in 60a97e9366373772efbd00368eaa7a4518ea2414 to the flat/unnested approach in #267 
See commit message.

Now I'll make this "mergeable" again.

**Comment by cburkert** [comment]
By the way, this PR is ready for merging again.

**Comment by jan-schreib** [comment]
CALM DOWN WE ARE LOOKING AT THIS!!1!

**Comment by thechauffeur** [comment]
@jan-schreib yeah well... did we? I mean we did but somehow we did not respond here -- why? In any case this should be reviewed and then finally merged :D

**Comment by thechauffeur** [comment]
> By the way, this PR is ready for merging again.

Well, not anymore :P

**Comment by jan-schreib** [comment]
> Well, not anymore :P

@cburkert U mad, bro?

File: src/test/java/de/qabel/core/config/ContactTestFactory.java [hunk 5]
```diff
@@ -20,8 +20,9 @@
...
 	@Override
 	public Contact create() {
-		Contact contact = new Contact(idFactory.create());
-		contact.setPrimaryPublicKey(qppkFactory.create());
-		return contact;
+		return new Contact(
+				idFactory.create(),
+				null,
+				qppkFactory.create());
 	}
```
Comment (thechauffeur) [inline src/test/java/de/qabel/core/config/ContactTestFactory.java]: No need to wrap lines here, you know.

File: src/test/java/de/qabel/core/drop/DropControllerTest.java [hunk 6]
```diff
@@ -16,142 +16,99 @@
...
-
-        recipientContacts = new Contacts();
-        recipientContacts.add(recipientContact);
-
-        dropController = new DropController();
-		
-	}
+    
+    @Before
+    public void setup() throws MalformedURLException, QblDropInvalidURL, InvalidKeyException {
```
Comment (thechauffeur) [inline src/test/java/de/qabel/core/drop/DropControllerTest.java]: These two lines have not really changed. Please do not just change the whitespaces in lines you don't touch.

File: src/main/java/de/qabel/core/drop/DropMessage.java [hunk 7]
```diff
@@ -1,80 +1,102 @@
...
     }
 
-    public void setData(T data) {
-        this.data = data;
+    /**
+     * Enable/disable acknowledge request for this drop message.
+     *
+     * @param enabled
+     */
+    public void enableAcknowledgeing(boolean enabled) {
```
Comment (thechauffeur) [inline src/main/java/de/qabel/core/drop/DropMessage.java]: Since the default is `enabled` this name is misleading.

Also it does not check whether the caller already had enabled acknowledging -- and this already generated an acknowledge id -- by using the normal constructor almost everyone would use (see above).

File: src/main/java/de/qabel/core/drop/DropMessage.java [hunk 8]
```diff
@@ -1,80 +1,102 @@
...
-        setData(data);
-    }
-
-    public int getVersion() {
-        return version;
-    }
-
-    public void setVersion(int version) {
-        this.version = version;
+    public DropMessage(Entity sender, T data) {
```
Comment (thechauffeur) [inline src/main/java/de/qabel/core/drop/DropMessage.java]: This magically enables acknowledging which is not even documented. I'm also not sure if this should be the default. Also see comment [below](https://github.com/Qabel/qabel-core/pull/269/files#r24431178).

Maybe we should change this constructor and add a third needed parameter `boolean enableAckowledging`.

**Comment by thechauffeur** [comment]
Besides my comments @jan-schreib and I ACK this PR. Also a new merge with master would be nice.

**Comment by jan-schreib** [comment]
I think its more sane to disable acknowledgement by default. Maybe add a new constructor with a 3rd parameter, but I don't like that idea much. Needs further discussion.

File: src/main/java/de/qabel/core/drop/DropMessage.java (hunk 8 above)
Comment (jan-schreib) [inline src/main/java/de/qabel/core/drop/DropMessage.java]: I think its more sane to disable acknowledgement by default. Maybe add a new constructor with a 3rd parameter, but I don't like that idea much. Needs further discussion.

File: src/test/java/de/qabel/core/drop/DropControllerTest.java (hunk 6 above)
Comment (cburkert) [inline src/test/java/de/qabel/core/drop/DropControllerTest.java]: Oh, that must be a bad relict of my conflict resolution. I guess you don't want me to change it back, do you?

File: src/main/java/de/qabel/core/drop/DropMessage.java (hunk 8 above)
Comment (cburkert) [inline src/main/java/de/qabel/core/drop/DropMessage.java]: Ok, I'll make no-acknowledging the new default. I'd say, in `send` one should then call

    dropMessage.enableAcknowledgeing(true);

and off it goes.
No need for changing constructors then.

File: src/main/java/de/qabel/core/drop/DropMessage.java (hunk 7 above)
Comment (cburkert) [inline src/main/java/de/qabel/core/drop/DropMessage.java]: I'll prevent re-generation of ackIds if they are already set. But if NOACK will be be the new default, this function should no longer be misleading, so I can keep the rest as it is.

**Comment by cburkert** [comment]
PR is ready for merging again.

File: src/test/java/de/qabel/core/drop/DropControllerTest.java (hunk 6 above)
Comment (thechauffeur) [inline src/test/java/de/qabel/core/drop/DropControllerTest.java]: No, but just because you asked so nice :P

File: src/main/java/de/qabel/core/drop/DropMessage.java (hunk 7 above)
Comment (thechauffeur) [inline src/main/java/de/qabel/core/drop/DropMessage.java]: Aye.

File: src/main/java/de/qabel/core/drop/DropMessage.java (hunk 8 above)
Comment (thechauffeur) [inline src/main/java/de/qabel/core/drop/DropMessage.java]: Aye.

File: src/main/java/de/qabel/core/drop/DropMessage.java [hunk 9]
```diff
@@ -1,80 +1,107 @@
...
-        this.data = data;
+    /**
+     * Enable/disable acknowledge request for this drop message.
+     * Acknowledging is disabled by default.
+     *
+     * @param enabled true enables acknowledging
+     */
+    public void enableAcknowledgeing(boolean enabled) {
+    	if (enabled && this.acknowledgeId != NOACK) {
+    		this.acknowledgeId = generateAcknowledgeId();
```
Comment (thechauffeur) [inline src/main/java/de/qabel/core/drop/DropMessage.java]: This will overwrite an existing acknowledgeId **and** will render it useless because acknowledging can never be enabled once NOACK has been set.

File: src/main/java/de/qabel/core/drop/DropMessage.java (hunk 8 above)
Comment (thechauffeur) [inline src/main/java/de/qabel/core/drop/DropMessage.java]: BTW, do you plan to implement this in this PR? If so we could use a (some) test(s) `DropControllerTest` for `send` and test the new acknowledgement true/false behaviour -- which could later be extended to test acknowledging `DropMessages`.

**Comment by thechauffeur** [comment]
@cburkert if you do a `git show 7fb8fb7bb7fa76c3f4d8154a17c9c15c62b4a06e` you can see -- again -- a lot of white space only changes in the diff. I don't know how you are able to introduce these even in a simple merge commit. Maybe you should use things like the git console or git gui tools like `gitk` more often instead of relying on your IDE :P

I would merge this PR anyway (if not for the bug mentioned above), nevertheless this is somewhat annoying (of course it is not really your fault though -- it is the mix of tabs and spaces).

File: src/main/java/de/qabel/core/drop/DropMessage.java (hunk 9 above)
Comment (cburkert) [inline src/main/java/de/qabel/core/drop/DropMessage.java]: Oh, stupid me. I'll fix it and write a test.

Comment (thechauffeur) [inline src/main/java/de/qabel/core/drop/DropMessage.java]: :+1:

File: src/main/java/de/qabel/core/drop/DropMessage.java (hunk 8 above)
Comment (cburkert) [inline src/main/java/de/qabel/core/drop/DropMessage.java]: No, I didn't plan to implemented acknowledged sending in this PR. I'm putting the acknowledge enabling tests into a dedicated test without dependencies to DropController.

**Comment by cburkert** [comment]
@thechauffeur Believe me, I don't use my IDE for git stuff at all. The reason for those trailing white spaces being removed is as simple as stupid: I rewrote almost the entire Identity class during the introduction of the super class and discarded some trailing white spaces around my modifications. Conflicting changes by @f-porter in the same file caused git automatic conflict resolution to take my stripped version.
Long story short: I didn't pick those changes manually for the merge commit but after auto-resolving it really looks stupid, without any doubt. I promise to take care of this in the future...

File: src/main/java/de/qabel/core/drop/DropMessage.java (hunk 9 above)
Comment (cburkert) [inline src/main/java/de/qabel/core/drop/DropMessage.java]: Fixed.

**Comment by jan-schreib** [comment]
I don't care about the whitespaces right now. You fixed the last bug and this is good to go. Merge.
